sw/device/tests/dif/dif_padctrl_test.cc

```diff
@@ -0,0 +1,481 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+//
+#include <set>
+#include <vector>
+
+#include "gtest/gtest.h"
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/testing/mock_mmio.h"
+
+#include "sw/device/lib/dif/dif_padctrl.h"
+#include "padctrl_regs.h"  // Generated.
+
+namespace dif_padctrl_test {
+namespace {
+
+class DifPadctrlTest : public testing::Test, public mock_mmio::MmioTest {
+ protected:
+  /**
+   *
+   * Attribute field informtion for a pad (MIO or DIO)
+   */
+  struct AttrField {
+    ptrdiff_t reg_offset; /**< register offset for register holding field. */
+    uintptr_t bit_mask;   /**< bit mask of field. */
+    uintptr_t bit_offset; /**< offset of field within register. */
+    uint32_t pad_num;     /**< which pad it corresponds to. */
+  };
+
+  // Setup attribute field info for all DIO and MIO pads
+  std::vector<AttrField> dio_attr_fields_ = {
```

Comment:
You probably could do something what I have done in pinmux to make these vectors smaller. You have auto generated field properties (width in bits, etc...), so given the register offset and last field offset in the register, it is easy to crawl every field in the register.

That's what I do in pinmux:

```
  struct Register {
    ptrdiff_t offset;    // Register offset from the base.
    uint8_t last_field;  // Last field offset in the register.
  };

  // Set multireg expectations, one field per call for every multireg register.
  void ExpectSetTests(const std::vector<Register> &regs, uint32_t field_width,
                      uint32_t field_mask, uint32_t value) {
    for (const auto &reg : regs) {
      for (uint32_t i = 0; i <= reg.last_field; i += field_width) {
        EXPECT_MASK32(reg.offset, {{i, field_mask, value}});
      }
    }
  }
```

Reply:
So instead of having to set 31 values by hand, I only need to do 7.

Reply:
I was intentionally going for the most simple/stupid method to get the relevant field info, to avoid tripping up over some issue in the code that generates the info, this guarantees we're using the same field info reggen put out. I've added a comment to explain the reasoning.

Reply:
I would ++ the "most simple/stupid method". Test code should prefer robustness over otherwise repeating itself; DRY doesn't apply IMO.

Though I would not object to a simple token-pasting macro to make the code less onerous to write out.

Reply:
I agree with you guys, I just feel like what I have suggested is a good trade-off between being repetitive and simplicity of the test cases. It's only a simple for loop.

It is matter of personal preference, I think we shouldn't bikeshed unittests too much.